### ISSUE TITLE
Directly using heapq as priority queue in ShortestPathChallenge breaks the specification

### ISSUE THREAD

**Issue by AGrondin** [csgames/cscoins#30]
The documentation states clearly:
"Find the shortest path. If there's multiple possibility, use the first one, and we check the nearest neighbors in that order : down, up, right, left."

I believe you tried to implement this ordering by inserting the positions in order:

def neighbors(self, pos):
        (row, col) = pos
        results = [(row+1, col), (row-1, col), (row, col+1), (row, col-1)]

For the priority queue, you used a heapq. In section 8.5.2 of the [heapq page](https://docs.python.org/3/library/heapq.html), they explain how to use a heapq to build a priority queue. This is not the way ShortestPathChallenge.py builds the heap , resulting in an inherently unstable implementation. 

You can easily see this by changing the order of neighbour insertion, and noticing that it won't have an effect on the final path. This is because insertion order does not impact the position of the element in the heap. The ordering will be based first on priority, then on the first and second elements of the tuple. 

Since heapq acts as a min-heap, [(row+1, col), (row-1, col), (row, col+1), (row, col-1)], if they have the same priority, will be popped as [(row-1,col),(row,col-1),(row,col+1),(row+1,col)]. 

So instead of looking at your neighbours in order down,up,right,left, you will instead look at them in order up, left, right, down. 

In the file I've included, you will find a comparison of two paths, one generated according to your documentation, and the other generated by your software.

[generated_paths.txt](https://github.com/csgames/cscoins/files/867156/generated_paths.txt)

You can use your code to generate this path using last_solution_hash
"215e7dde782ea4c3134f693eb72522f0aafce006afeb50ccd3317724cd4920e5"
and nonce 13349735025080016891. 

We can clearly see the issue. While both our paths have the same length, yours goes left instead of down, when you specification says it shouldn't. 

The bug results from the fact that you most likely copy-pasted a dijkstra implementation, probably from section 1.3.3 of this [this](http://www.redblobgames.com/pathfinding/a-star/implementation.html) blog, without taking the time to understand the libraries you were using.

**Comment by antifob** [comment]
Fix it, that's it. Many of us have implemented solvers weeks ago.
Be glad you found a way and implement a solver now.

**Comment by AGrondin** [comment]
Oh sorry, it must have come across the wrong way. I was an easy fix for me,
and, I mean shit, the games are getting started.

But the official specs were and are still incorrect, though, and I guess
last night I really wanted you to know exactly why.

Good luck with maintenance during the games, I hope everything goes well!

**Comment by jordsti** [comment]
> But the official specs were and are still incorrect, though, and I guess
last night I really wanted you to know exactly why.

I'm taking at look at this right now.

**Comment by AGrondin** [comment]
Yeah, but please don't fix it. I was just pointing out the inconsistency.

**Comment by jordsti** [comment]
I mean, I won't change the code, but only the specs to fix the inconsistency.
